src/fs/serve_file.rs

```diff
@@ -0,0 +1,83 @@
+use crate::log;
+use crate::{Body, Endpoint, Request, Response, Result, StatusCode};
+use std::io;
+use std::path::Path;
+
+use async_std::path::PathBuf as AsyncPathBuf;
+use async_trait::async_trait;
+
+pub struct ServeFile {
+    path: AsyncPathBuf,
+}
+
+impl ServeFile {
+    /// Create a new instance of `ServeFile`.
+    pub(crate) fn init(path: impl AsRef<Path>) -> io::Result<Self> {
+        let file = path.as_ref().to_owned().canonicalize()?;
+        Ok(Self {
+            path: AsyncPathBuf::from(file),
+        })
+    }
+}
+
+#[async_trait]
+impl<State: Clone + Send + Sync + 'static> Endpoint<State> for ServeFile {
+    async fn call(&self, _: Request<State>) -> Result {
+        if !self.path.exists().await {
```

Comment:
This looks like a potential [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) issue. I think we should just try to read the file immediately and return `StatusCode::NotFound` if that fails with `io::ErrorKind::NotFound`.

`serve_dir()` has the same issue. We could also land this as is and fix it in a separate PR, whichever you prefer

Reply:
Ah yes, I copied this from `serve_dir` I'd prefer to fix this as part of this PR, at least for `serve_file` I don't really want to add known bugs to the code-base if I can avoid it. I'll fix this later today.

Reply:
Fixed